### PR TITLE
Threadpool isolation for MDC

### PR DESCRIPTION
**Pull request by bbrehm**
Related: https://github.com/ShiftLeftSecurity/common/pull/126 https://github.com/ShiftLeftSecurity/codepropertygraph/pull/1527 slack #mdc-propagation https://github.com/ShiftLeftSecurity/product/issues/8624 https://github.com/ShiftLeftSecurity/codescience/pull/5602

Theoretically this wants to be its own minimal repo, but that comes with red tape.

### CODE REVIEW

File: threadpool-isolation/src/main/java/io/shiftleft/threadpoolisolation/StructuredConcurrencyContext.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    private static volatile ExecutorService globalExecutor = null;
+
+    public static StructuredConcurrencyContext getCurrent() {
+        return currentContext.get();
+    }
+
+    public static ForkJoinPool getForkJoinPool() {
+        StructuredConcurrencyContext ctx = getCurrent();
+        if (ctx != null) return ctx.forkJoinPool();
+        else if (ctx == null && !returnDefaultOnMissing) return null;
```
Comment (mpollmeier) [inline threadpool-isolation/src/main/java/io/shiftleft/threadpoolisolation/StructuredConcurrencyContext.java:18]:
```suggestion
        else if (!returnDefaultOnMissing) return null;
```

File: threadpool-isolation/src/main/java/io/shiftleft/threadpoolisolation/StructuredConcurrencyContext.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+                    globalExecutor = Executors.newCachedThreadPool();
+                return globalExecutor;
+            }
+        }
+    }
+
+    public static ThreadFactory getThreadFactory() {
+        StructuredConcurrencyContext ctx = getCurrent();
+        if (ctx != null) return ctx.threadFactory();
+        else if (ctx == null && !returnDefaultOnMissing) return null;
```
Comment (mpollmeier) [inline threadpool-isolation/src/main/java/io/shiftleft/threadpoolisolation/StructuredConcurrencyContext.java:49]:
```suggestion
        else if (!returnDefaultOnMissing) return null;
```

File: threadpool-isolation/src/main/java/io/shiftleft/threadpoolisolation/StructuredConcurrencyContext.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
+package io.shiftleft.threadpoolisolation;
+
+import java.util.concurrent.*;
+
+abstract public class StructuredConcurrencyContext implements AutoCloseable {
```
Comment (mpollmeier) [inline threadpool-isolation/src/main/java/io/shiftleft/threadpoolisolation/StructuredConcurrencyContext.java:5]: Can you please add some docs for the public api, e.g. a brief "what this class does", differences between enter and enterOnce, usage examples.

File: threadpool-isolation/src/main/java/io/shiftleft/threadpoolisolation/StructuredConcurrencyContext.java [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+            currentContext.set(this);
+            r.run();
+        } finally {
+            currentContext.set(oldCtx);
+            exitHook(unrelatedContexts);
+        }
+    }
+
+    public void shutdownPools() {
+        this.forkJoinPool().shutdown();
```
Comment (mpollmeier) [inline threadpool-isolation/src/main/java/io/shiftleft/threadpoolisolation/StructuredConcurrencyContext.java:77]: can these potentially fail, i.e. do we need to guard these with try/arm?

**Review by mpollmeier — COMMENTED** [review]
Apart from the comments on the files - can you please add some tests?

**Comment by bbrehm** [comment]
Superseded.

**Pull request closed without merging**